### DISCUSSION TITLE
Specifying a JSON enum string converter for a list property

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/runtime#99777] [General] [answered]
I have a property which is an enumeration of enums (in my case it is a `HashSet<T>` but this minimal repro is representative), and I am unsure how to specify the JSON enum string converter:

``` csharp
using System.Text.Json;
using System.Text.Json.Serialization;

using NUnit.Framework;

const string json = """
{
    "values": ["One", "Two", "Three"]
}
""";

HasNumbers? hasNumbers = JsonSerializer.Deserialize<HasNumbers>(json, NumberSerializerContext.Default.HasNumbers);
Assert.That(hasNumbers, Is.Not.Null);
Number[]? numbers = hasNumbers.Values;

Assert.That(numbers, Is.Not.Null);
Assert.That(numbers, Has.Length.EqualTo(3));
Assert.Multiple(() =>
{
    Assert.That(numbers[0], Is.EqualTo(Number.One));
    Assert.That(numbers[1], Is.EqualTo(Number.Two));
    Assert.That(numbers[2], Is.EqualTo(Number.Three));
});

[JsonSourceGenerationOptions(
// UseStringEnumConverter = true // this approach does not work
)]
[JsonSerializable(typeof(HasNumbers))]
public sealed partial class NumberSerializerContext : JsonSerializerContext;

public class HasNumbers
{
    // [JsonConverter(typeof(JsonStringEnumConverter<Number>))] // this does not work either
    public Number[]? Values { get; set; }
}

public enum Number
{
    One,
    Two,
    Three,
}
```

Is there any way to do what I am doing? Preferably in an AOT-friendly way.

**Comment by elgonzo** [comment]
Setting `UseStringEnumConverter = true` in the `[JsonSourceGenerationOptions]` attribute is the correct approach. However, this is not the only setting you need to take care of here.

Note that `Values` != `values`. To make deserialization work as desired, you also need to enable case-insensitive property name handling:
```C#
[JsonSourceGenerationOptions(
  UseStringEnumConverter = true,
  PropertyNameCaseInsensitive = true
)]
[JsonSerializable(typeof(HasNumbers))]
public sealed partial class NumberSerializerContext : JsonSerializerContext;
```

**Reply by RenderMichael** [reply]
Thanks for the tip, that got it working. Careless mistake on my part.

However, using the context-wide string enum converter property was a last-ditch attempt on my part. Is there a more localized way of pulling this off, ideally on the `Values` property itself?

**Reply by RenderMichael** [reply]
The `Number` enum (or rather, the actual enum I'm working with outside the minimal repro) is from a dependency, outside of my control.

**Reply by elgonzo** [reply, marked as answer]
> Is there a more localized way of pulling this off, ideally on the Values property itself?

Depends on whether you want the desired behavior specifically only on the `Values` property, or specifically on the `Numbers` enum type regardless where it is being used in the object graph being (de)serialized.

If the behavior should be specific to the `Values` property (with other properties of the same enum type being unaffected), currently you would need to write a custom JsonConverter that then executes the StringEnumConverter on the collection items, similar to this implementation: https://github.com/dotnet/runtime/issues/54189#issuecomment-861628532. This converter would then be applied to the `Values` property using the `[JsonConverter]` attribute. There are open issues concerning the inability to easily specify JsonConverters for collections items (such as this: https://github.com/dotnet/runtime/issues/54189, for example), but as far as i am aware addressing this has been repeatedly pushed into the future by the team... :-(

If the StringEnumConverter should be applied to the `Numbers` type wherever it is being used, the solution is much simpler. Change the `[JsonSourceGenerationOptions]` annotation to:
```C#
[JsonSourceGenerationOptions(
  Converters = new[] { typeof(JsonStringEnumConverter<Number>) },
  PropertyNameCaseInsensitive = true
)]
[JsonSerializable(typeof(HasNumbers))]
public sealed partial class NumberSerializerContext : JsonSerializerContext;
```
